### PR TITLE
Universal sierra -> casm compilation library

### LINKED ISSUE #1211: Implement rust library for unversal compilation of `Sierra` -> `CASM`
It should be a separate package, which allows the user to provide any sierra,
and it will automatically detect & compile it into CASM

### PR DESCRIPTION
**Pull request by war-in**
Closes #1211

## Introduced changes

- new sierra-casm crate
- sierra-casm exposes the `compile` function which takes `FlattenedSierraClass` from starknet-rs and returns `CasmContractClass`

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/cheatnet/src/forking/state.rs
Comment (MaksymilianDemitraszek) [inline crates/cheatnet/src/forking/state.rs:1]: Can we do migration in a separate PR? We need to make sure we removed all related structures and code.

File: crates/sierra-casm/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+use cairo_lang_starknet::contract_class::{ContractClass, ContractEntryPoints};
+use cairo_lang_starknet_v1_0_0_alpha6::casm_contract_class::CasmContractClass as CasmContractClassV1Alpha;
+use cairo_lang_starknet_v1_0_0_alpha6::contract_class::ContractClass as ContractClassV1Alpha;
+use cairo_lang_starknet_v1_0_0_rc0::casm_contract_class::CasmContractClass as CasmContractClassV1Rc0;
+use cairo_lang_starknet_v1_0_0_rc0::contract_class::ContractClass as ContractClassV1Rc0;
+use cairo_lang_starknet_v1_1_1::casm_contract_class::CasmContractClass as CasmContractClassV1_1;
+use cairo_lang_starknet_v1_1_1::contract_class::ContractClass as ContractClassV1_1;
+
+pub fn compile(definition: &FlattenedSierraClass) -> anyhow::Result<CasmContractClass> {
+    macro_rules! compile_contract {
```
Comment (MaksymilianDemitraszek) [inline crates/sierra-casm/src/lib.rs:18]: This macro is not very readable, this is imo not a good use case, can we just write code instead?

File: crates/sierra-casm/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    let json = serde_json::json!({
+        "sierra_program": converted_sierra_program,
+        "contract_class_version": value.contract_class_version,
+        "entry_points_by_type": converted_entry_points,
+    });
+    serde_json::from_value::<T>(json)
+}
+
+fn casm_to_main_casm<T>(value: &T) -> anyhow::Result<CasmContractClass>
```
Comment (MaksymilianDemitraszek) [inline crates/sierra-casm/src/lib.rs:57]: I don't undestand what this function does.

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
Overall LGTM, but please migrate forge in a separate PR, we need to make sure we remove all related code.

File: crates/sierra-casm/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }};
+    }
+
+    compile_contract!(ContractClass, CasmContractClass);
+    compile_contract!(ContractClassV1_1, CasmContractClassV1_1);
+    compile_contract!(ContractClassV1Rc0, CasmContractClassV1Rc0);
+    compile_contract!(ContractClassV1Alpha, CasmContractClassV1Alpha);
+
+    unreachable!()
```
Comment (Arcticae) [inline crates/sierra-casm/src/lib.rs:33]: I think we settled on not using the unreachable macro where it can be a panic (i think this case can even be a handleable error since we are extracting it out to a crate)

File: crates/sierra-casm/src/lib.rs (hunk 1 above)
Comment (Arcticae) [reply crates/sierra-casm/src/lib.rs:18]: Perhaps a regular function would be better

File: crates/sierra-casm/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+use cairo_lang_starknet::casm_contract_class::CasmContractClass;
+use cairo_lang_starknet::contract_class::{ContractClass, ContractEntryPoints};
+use cairo_lang_starknet_v1_0_0_alpha6::casm_contract_class::CasmContractClass as CasmContractClassV1Alpha;
+use cairo_lang_starknet_v1_0_0_alpha6::contract_class::ContractClass as ContractClassV1Alpha;
+use cairo_lang_starknet_v1_0_0_rc0::casm_contract_class::CasmContractClass as CasmContractClassV1Rc0;
+use cairo_lang_starknet_v1_0_0_rc0::contract_class::ContractClass as ContractClassV1Rc0;
+use cairo_lang_starknet_v1_1_1::casm_contract_class::CasmContractClass as CasmContractClassV1_1;
+use cairo_lang_starknet_v1_1_1::contract_class::ContractClass as ContractClassV1_1;
+
+pub fn compile(definition: &FlattenedSierraClass) -> anyhow::Result<CasmContractClass> {
```
Comment (Arcticae) [inline crates/sierra-casm/src/lib.rs:17]: I think the interface should be more universal, so that the user does not have to use `starknet::core` in order to compile

File: crates/sierra-casm/src/lib.rs (hunk 2 above)
Comment (Arcticae) [reply crates/sierra-casm/src/lib.rs:57]: Netiher do i 😬

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: crates/sierra-casm/src/lib.rs (hunk 2 above)
Comment (war-in) [reply crates/sierra-casm/src/lib.rs:57]: We do have to convert `CasmContractClass` which we got from the compilation to the latest CasmContractClas. Will add a description and maybe rename this function

File: crates/cheatnet/src/forking/state.rs
Comment (Arcticae) [inline crates/cheatnet/src/forking/state.rs:1]: Changes from this file should be reverted

File: crates/sierra-casm/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            }
+        }};
+    }
+
+    compile_contract!(ContractClass, CasmContractClass);
+    compile_contract!(ContractClassV1_1, CasmContractClassV1_1);
+    compile_contract!(ContractClassV1Rc0, CasmContractClassV1Rc0);
+    compile_contract!(ContractClassV1Alpha, CasmContractClassV1Alpha);
+
+    Err("Unable to compile Sierra to Casm".to_string())
```
Comment (Arcticae) [inline crates/sierra-casm/src/lib.rs:35]: A reason would be nice here

File: crates/sierra-casm/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+/// converts `CasmContractClass` from the old `cairo_lang_starknet` library version
+/// to the `CasmContractClass` from the newest version
+fn old_casm_to_newest_casm<T>(value: &T) -> CasmContractClass
+where
+    T: Serialize + DeserializeOwned,
+{
+    let serialized = serde_json::to_value(value).unwrap();
+    serde_json::from_value::<CasmContractClass>(serialized).unwrap()
```
Comment (Arcticae) [inline crates/sierra-casm/src/lib.rs:45]: We should probably chain those errors

Comment (Arcticae) [inline crates/sierra-casm/src/lib.rs:45]: Shouldn't this be of type `T`?

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: crates/sierra-casm/src/lib.rs (hunk 6 above)
Comment (war-in) [reply crates/sierra-casm/src/lib.rs:45]: I think it is not necessary because Casm should not change. This serialization is only needed because of the types

Comment (war-in) [reply crates/sierra-casm/src/lib.rs:45]: Nope, we want to deserialize it to the latest CasmContractClass

File: crates/sierra-casm/src/lib.rs (hunk 1 above)
Comment (war-in) [reply crates/sierra-casm/src/lib.rs:20]: 🔒

File: crates/sierra-casm/src/lib.rs (hunk 6 above)
Comment (Arcticae) [reply crates/sierra-casm/src/lib.rs:45]: Ah okay

**Review by Arcticae — APPROVED** [review]
